effectful/handlers/pyro.py

```diff
@@ -252,190 +265,264 @@ def __repr__(self):
+    def op(self):
+        return call
 
-        # shape: | sample_shape | named | batch_shape | event_shape |
-        pos_tensor_r = torch.permute(pos_tensor, sample_dims + named_dims + base_dims)
+    @property
+    def args(self):
+        return self._args
 
-        return pos_tensor_r
+    @property
+    def kwargs(self):
+        return self._kwargs
 
-    def _from_positional(self, value: torch.Tensor) -> torch.Tensor:
-        # maximal value shape: | sample_shape | named | batch_shape | event_shape |
-        return self.naming.apply(value)
+    @property
+    def _base_dist(self):
+        return self.args[0](*self.args[1:])
 
     @property
     def has_rsample(self):
-        return self.base_dist.has_rsample
+        return self._base_dist.has_rsample
 
     @property
     def batch_shape(self):
-        return (
-            torch.Size([s for s in self.indices.values()]) + self.base_dist.batch_shape
-        )
+        return self._base_dist.batch_shape
 
     @property
     def event_shape(self):
-        return self.base_dist.event_shape
+        return self._base_dist.event_shape
 
     @property
     def has_enumerate_support(self):
-        return self.base_dist.has_enumerate_support
+        return self._base_dist.has_enumerate_support
 
     @property
     def arg_constraints(self):
-        return self.base_dist.arg_constraints
+        return self._base_dist.arg_constraints
 
     @property
     def support(self):
-        return self.base_dist.support
-
-    def __repr__(self):
-        return f"PositionalDistribution({self.base_dist})"
+        return self._base_dist.support
 
     def sample(self, sample_shape=torch.Size()):
-        return self._to_positional(self.base_dist.sample(sample_shape))
+        return self._base_dist.sample(sample_shape)
 
     def rsample(self, sample_shape=torch.Size()):
-        return self._to_positional(self.base_dist.rsample(sample_shape))
+        return self._base_dist.rsample(sample_shape)
 
     def log_prob(self, value):
-        return self._to_positional(
-            self.base_dist.log_prob(self._from_positional(value))
-        )
+        return self._base_dist.log_prob(value)
 
     def enumerate_support(self, expand=True):
-        return self._to_positional(self.base_dist.enumerate_support(expand))
+        return self._base_dist.enumerate_support(expand)
 
+    def __getitem__(self, key: Collection[Operation[[], int]]):
+        return named_distribution(self, *key)
 
-class NamedDistribution(pyro.distributions.torch_distribution.TorchDistribution):
-    """A distribution wrapper that lazily names leftmost dimensions."""
 
-    def __init__(
-        self,
-        base_dist: pyro.distributions.torch_distribution.TorchDistribution,
-        names: Collection[Operation[[], int]],
-    ):
-        """
-        :param base_dist: A distribution with batch dimensions.
+@defterm.register(TorchDistribution)
+@defterm.register(TorchDistributionMixin)
+@functools.singledispatch
```

Comment:
Why the second layer of `singledispatch` here? Can we just use `defterm.register` and `defdata.register` directly? That way it's easier to add support for new distributions defined outside the library by adding `defterm` patterns in situ.